### ISSUE TITLE
New wazuh-indexer build including elasticsearch-oss and Opendistro for Elasticsearch research

### ISSUE THREAD

**Issue by alberpilot** [wazuh/wazuh-packages#666]
The build `wazuh-indexer` will contains the `elasticsearch-oss` and all open distro software. To be taken into account: 

- An `elasticsearch-oss` build under `/usr/share/elasticsearch/`. 
- The `/etc/elasticsearch` directory will contain both `elasticsearch.yml` and a folder (`certs`) dedicated to storing certificates used to encrypt the communications with others `elasticsearch` nodes and `wazuh-manager`s. 
- `elasticsearch.yml` will contain our custom unified configuration. 
- Our custom configurations and tools are deployed (password-changer.sh, certificates-tools, etc,)
- Opendistro for Elasticsearch plugins files are deployed. The elasticsearch build must be done with the plugins included. 
- All non-necessaries folders must be removed. (To be defined)
- Create an elasticsearch service like the default installation.
- Create elasticsearch user and group. 

The obtained build will be used in later issues for defining SPECs and creating packages.

**Comment by sbruno** [comment]
I've being working on generating a RPM with the requested contents.

There is available both an [elasticsearch-oss-7-10-2](https://www.elastic.co/downloads/past-releases/elasticsearch-oss-7-10-2) tar.gz containing a jdk and [elasticsearch-oss-no-jdk-7-10-2](https://www.elastic.co/downloads/past-releases/elasticsearch-oss-no-jdk-7-10-2) which does not include it.

I uploaded both to S3 so the SPEC can base on them:
https://packages-dev.wazuh.com/deps/elasticsearch-oss-7.10.2-no-jdk-linux-x86_64.tar.gz
https://packages-dev.wazuh.com/deps/elasticsearch-oss-7.10.2-linux-x86_64.tar.gz

The problem with the one that does not include the JDK is that it will show a warning on every command:
![image](https://user-images.githubusercontent.com/39304/111413877-cfe89800-86bd-11eb-8828-5817e84cab05.png)
But that could be removed editing the `elasticsearch-env` file.

However the one that includes JDK includes OpenJDK version 15 which may be a good thing:
![image](https://user-images.githubusercontent.com/39304/111414004-058d8100-86be-11eb-95d7-99c25d10191e.png)

For the moment I'm using this one until I get a working elasticsearch installation from the RPM so I don't have to care about setting up Java.

I'm building the RPM from a SPEC locally so I can develop and test faster, I have the elasticsearch tar.gz files locally.

For the plugins a similar approach can be done:

I downloaded the zip files of most of them from the URLs specified here: https://opendistro.github.io/for-elasticsearch-docs/docs/install/plugins/

Installing them seems to be just unzipping them to the plugins folder (I need to check that they work though). Except for a couple:

- opendistro-performance-analyzer needs some manual configuration
- k-NN is only available as part of the all-in-one installs: Docker, RPM, and Debian.

This seems to be because that plugins has C++ code so it is built for different architectures.
It may make sense to leave this plugin alone for manual installation instead of having wazuh-indexer builds for different architectures, since it is mostly Java, but it is something to define.

The path to the configuration which is now `/etc/wazuh-indexer` is defined in `ES_PATH_CONF` environment variable, which is set by `elasticsearch-env`.

In the default (tar.gz) installation that we are using, it defaults `/config` in the elasticsearch installation path, so it would be `/usr/share/wazuh-indexer/config`, and that failed.
In official RPM builds it sources `/etc/sysconfig/elasticsearch` and in DEB `/etc/default/elasticsearch`

This is correspondingly set by the elasticsearch build process. The file has a [placeholder](https://github.com/elastic/elasticsearch/blob/master/distribution/src/bin/elasticsearch-env#L87) that is then set depending on the package type in their [build.gradle](https://github.com/elastic/elasticsearch/blob/0db220f5a175a9b547f98e4cb2e5676604ce2444/distribution/build.gradle#L528)

```
      'source.path.env': [
        'deb': 'source /etc/default/elasticsearch',
        'rpm': 'source /etc/sysconfig/elasticsearch',
        'def': 'if [ -z "$ES_PATH_CONF" ]; then ES_PATH_CONF="$ES_HOME"/config; fi',
      ],
```

At the moment in the SPEC what I'm doing is replacing `if [ -z "$ES_PATH_CONF" ]; then ES_PATH_CONF="$ES_HOME"/config; fi` by `source /etc/sysconfig/elasticsearch`, which seems to be quite hacky and fragile but it works.

I also used `rpmrebuild` on an existing installation of `elasticsearch-oss` and got a quite large SPEC file, which has the code for creating the elasticsearch user and group among other things. I copied all that to my SPEC replacing elasticsearch by wazuh-indexer and it installed and some commands work, like listing the plugins, but still need to work in getting the service to start.
![image](https://user-images.githubusercontent.com/39304/111416050-dbd65900-86c1-11eb-9bba-2cc5811c88f1.png)

It might be a good idea to investigate the build process used for generating the rpm and deb files by elasticsearch, because the SPEC generated by rpmrebuild is terribly long and changing manually elasticsearch for wazuh-indexer will make it difficult to maintain.

**Comment by sbruno** [comment]
I've been able to get all the ODFE plugins including KNN from its RPM file.

Using `rpmrebuild` I obtained the SPECs for `elasticsearch-oss` and all the plugins, and also extracted the contents of  `opendistro-knn-1.13.0.0-1.rpm` (plugin) and `opendistro-knnlib-1.13.0.0-1.rpm` (x86 library)

Using `rpmrebuild` on `elasticsearch-oss` I also obtained the scripts for setting up `elasticsearch` user and more

From `elasticsearch-oss` RPM I also extracted the `init.d,` `sysconfig` and other scripts used for starting the service.

I generated a SPEC combining al those, but faced problems.

The RPM installs but shows a `Failed to execute operation: No such file or directory` message so there must be some path error.

The service does not start when using `systemctl start elasticsearch.service`

Looking at the logs, there is some problem with the knn plugin. The `elasticsearch.log` file shows:

```
[2021-03-17T23:47:26,678][ERROR][o.e.b.Bootstrap          ] [node-1] Exception
java.lang.IllegalArgumentException: Values less than -1 bytes are not supported: -17311744b
	at org.elasticsearch.common.unit.ByteSizeValue.<init>(ByteSizeValue.java:79) ~[elasticsearch-7.10.2.jar:7.10.2]
	at com.amazon.opendistroforelasticsearch.knn.index.KNNSettings.parseknnMemoryCircuitBreakerValue(KNNSettings.java:337) ~[?:?]
```

Then tried removing the knn plugin and starting the server again, but got problems with the `opendistro-performance-analyzer`

```
[2021-03-17T23:57:53,197][WARN ][stderr                   ] [node-1] java.nio.file.NoSuchFileException: /dev/shm/performanceanalyzer/1616025465000.tmp
[2021-03-17T23:57:53,204][ERROR][c.a.o.e.p.r.EventLogFileHandler] [node-1] Error writing entry 'NOT_INITIALIZED'. Cause: /dev/shm/performanceanalyzer/1616025465000.tmp
```

These two plugins are exceptional because knn needs an OS dependent library and performance-analyzer needs extra configuration (apparently it needs its own service running and the RPM scripts I added probably don't set up everything it needs)

I even tried removing all plugins and putting in `/etc/elasticsearch/elasticsearch.yml` the default configuration (because wazuh's references the ODFE security plugin), and in that case the service still didn't start.
In that case, the call to `systemctl start elasticsearch.service` failed with timeout, but elasticsearch executable seemed to run. If while systemclt command was working I opened another console and executed `curl -S http://localhost:9200` I got the response

```
{
  "name" : "host1",
  "cluster_name" : "elasticsearch",
  "cluster_uuid" : "COUNwtKvQW642L5gwno3cA",
  "version" : {
    "number" : "7.10.2",
    "build_flavor" : "oss",
    "build_type" : "tar",
    "build_hash" : "747e1cc71def077253878a59143c1f785afa92b9",
    "build_date" : "2021-01-13T00:42:12.435326Z",
    "build_snapshot" : false,
    "lucene_version" : "8.7.0",
    "minimum_wire_compatibility_version" : "6.8.0",
    "minimum_index_compatibility_version" : "6.0.0-beta1"
  },
  "tagline" : "You Know, for Search"
}

```

But after the timeout the service stopped. In `elasticsearch.log` I saw:

```
[2021-03-18T01:32:33,634][INFO ][o.e.g.GatewayService     ] [host1] recovered [0] indices into cluster_state
[2021-03-18T01:32:33,638][INFO ][o.e.h.AbstractHttpServerTransport] [host1] publish_address {127.0.0.1:9200}, bound_addresses {[::1]:9200}, {127.0.0.1:9200}
[2021-03-18T01:32:33,638][INFO ][o.e.n.Node               ] [host1] started
[2021-03-18T01:33:32,088][INFO ][o.e.n.Node               ] [host1] stopping ...
[2021-03-18T01:33:32,097][INFO ][o.e.n.Node               ] [host1] stopped
[2021-03-18T01:33:32,097][INFO ][o.e.n.Node               ] [host1] closing ...
[2021-03-18T01:33:32,101][INFO ][o.e.n.Node               ] [host1] closed
```

So in summary the main problems at the moment for the RPM are:

- systemctl call not returning successfully
- plugin errors: knn, performance, maybe more.

**Comment by sbruno** [comment]
Now instead of generating an RPM I generated a tar.gz file containing the files added to a Centos 7 VM after a manual installation consisting of:

- unpacking elasticsearch-oss from tar.gz
- configuring it to start and checking it works.
  - The default configuration is in `/usr/share/elasticsearch/config`. Instead of using it, the config folder is moved to `/etc/elasticsearch` (`mv /usr/share/elasticsearch/config /etc/elasticsearch`
`chown elasticsearch:elasticsearch /etc/elasticsearch -R` ) and `ES_PATH_CONF` environment variable is set export `ES_PATH_CONF=/etc/elasticsearch` (then this is not needed because the path will be configured in a file that will be sourced by elasticsearch-env file)
- Installing [opendistro-security plugin](https://opendistro.github.io/for-elasticsearch-docs/docs/install/plugins/#security) and configuring it following the wazuh procedure found in all in one installation guide from [configuring elasticsearch](https://documentation.wazuh.com/current/installation-guide/open-distro/all-in-one-deployment/all_in_one.html#configuring-elasticsearch) section onwards and installing and configuring scripts. 
  -  This will generate certificates in `/etc/elasticsearch/certs`
- Installing the rest of the plugins following [the documentation](https://opendistro.github.io/for-elasticsearch-docs/docs/install/plugins/)
- Copying the `systemctl` related scripts and configuration files extracted from elasticsearch-oss RPM file to be able to start the service
- Editing `/usr/share/elasticsearch/bin/elasticsearch-env` to include `source /etc/sysconfig/elasticsearch` instead of `if [ -z "$ES_PATH_CONF" ]; then ES_PATH_CONF="$ES_HOME"/config; fi'` . This will cause that the proper configuration folder is set when using elasticsearch since `/etc/sysconfig/elasticsearch` will actually have `ES_PATH_CONF=/etc/elasticsearch`


After everything was working, packaged the files:

```
cd /

tar zcvf wazuh-indexer.tgz \
    usr/share/elasticsearch \
    etc/elasticsearch/ \
    etc/init.d/elasticsearch \
    etc/sysconfig/elasticsearch \
    usr/lib/sysctl.d/elasticsearch.conf \
    usr/lib/systemd/system/elasticsearch.service \
    usr/lib/tmpfiles.d/elasticsearch.conf \
    var/lib/elasticsearch/batch_metrics_enabled.conf \
    var/lib/elasticsearch/logging_enabled.conf \
    var/lib/elasticsearch/performance_analyzer_enabled.conf \
    var/lib/elasticsearch/rca_enabled.conf \
    var/log/elasticsearch/elasticsearch.log

```

The error I got with KNN plugin about a negative value on free memory went away when I increased the VM memory, it was set to just 1GB and I set it to 4. Performance analysis plugin also did not fail after executing the extra manual configuration

The problem that is still pending to solve is why the service seems not to be starting for systemctl causing it to timeout and kill it. Elasticsearch is actually functional but it looks like it does not return or is not set to run in background/daemon. Increasing the service timeout value allows to test elasticsearch while systemctl is waiting, for example

```
sed -i 's/TimeoutStartSec=75/TimeoutStartSec=9999/'  /usr/lib/systemd/system/elasticsearch.service
systemctl daemon-reload
systemctl start elasticsearch
```

**Comment by sbruno** [comment]
The problem with `systemctl start elasticsearch` never returning successfully even when elasticsearch started is because the tar.gz is not designed to be started this way, so it does not bundle the Java module that notifies and also the elasticsearch code verifies if the build type is `tar` and excludes the execution of that code.

These are related issues:
https://github.com/elastic/elasticsearch/issues/55477
https://github.com/elastic/cookbook-elasticsearch/issues/734

The changes done in the tar.gz to work were

- Editing `bin/elasticsearch-env`

```diff
ES_DISTRIBUTION_FLAVOR=oss
-ES_DISTRIBUTION_TYPE=tar
+ES_DISTRIBUTION_TYPE=rpm
ES_BUNDLED_JDK=true
```

This way the java command used for starting elasticsearch includes the parameter `-Des.distribution.type=rpm `making the code to execute.

- Adding the systemd module files which I extracted from the `elasticsearch-oss.rpm`:
```
/usr/share/elasticsearch/modules/systemd/plugin-descriptor.properties
/usr/share/elasticsearch/modules/systemd/plugin-security.policy
/usr/share/elasticsearch/modules/systemd/systemd-7.9.1.jar

```

This worked, but is definitely a hack. We are packaging the tar.gz and the applying many changes and additional files from the RPM bundle. There is some risk that the elasticsearch code has other conditionals like this that may fail and we have not tested yet.

It might be worth analyzing elasticsearch build process and see if we can adapt it to properly package the rpm and deb and add the ODFE plugins: https://github.com/elastic/elasticsearch/tree/v7.10.2/distribution

**Comment by sbruno** [comment]
I was able to generate a v2.tgz with the changes mentioned in the above comment and use it to set up elasticsearch and perform an all in one installation using a wazuh manager, filebeat and kibana in the same box.
Didn't find any problem at the moment.
I used this script below, which I located in the same folder where I had the tar.gz file:

```bash
# Add user
adduser elasticsearch

# Uncompress the tar.gz file and set permissions on var folders
tar zxvf wazuh-indexer.v2.tgz -C /
# Set directory permissions as done in RPM
#    %dir %attr(2750, elasticsearch, elasticsearch) "/var/lib/elasticsearch"
#    %dir %attr(2750, elasticsearch, elasticsearch) "/var/log/elasticsearch"
chown elasticsearch:elasticsearch /var/log/elasticsearch/ /var/lib/elasticsearch/
chmod 750 /var/log/elasticsearch/ /var/lib/elasticsearch/
chmod g+s /var/log/elasticsearch/ /var/lib/elasticsearch/

# Start the service
systemctl daemon-reload
systemctl enable elasticsearch
systemctl start elasticsearch

# Configure Java environment variables so it can be used by securityadmin tool
export JAVA_HOME=/usr/share/elasticsearch/jdk/
export PATH=/usr/share/elasticsearch/jdk/bin:$PATH

# Run securityadmin tool to initialize OpenDistro
/usr/share/elasticsearch/plugins/opendistro_security/tools/securityadmin.sh -cd /usr/share/elasticsearch/plugins/opendistro_security/securityconfig/ -nhnv -cacert /etc/elasticsearch/certs/root-ca.pem -cert /etc/elasticsearch/certs/admin.pem -key /etc/elasticsearch/certs/admin.key

# Execute test query to list plugins:
curl -XGET "https://localhost:9200/_cat/plugins?v=true&s=component&h=name,component,version,description" -k -u admin:admin

# Execute other queries to test some plugins:
curl -XGET "https://localhost:9200/_opendistro/_alerting/stats" -k -u admin:admin
curl -XGET "https://localhost:9200/_opendistro/_anomaly_detection/stats" -k -u admin:admin
curl -XGET "https://localhost:9200/_opendistro/_knn/node-1/stats" -k -u admin:admin

# This particular query fails, but it also failed in an environment with elasticsearch installed by RPM, it is something to investigate further. 
curl -XGET "https://localhost:9200/_opendistro/_performanceanalyzer/metrics?metrics=Latency,CPU_Utilization&agg=avg,max&dim=ShardID&nodes=all" -k -u admin:admin



# At this stage, elasticsearch is ready.
# The certificate used is for All-in-one installation It is valid for 127.0.0.1. Other certs should be made for distributed installation.
# But the all in one installation can be done now for testing, continuing by installing the manager, filebeat and Kibana, following Wazuh's documentation bug skipping elasticsearch section:
# https://documentation.wazuh.com/current/installation-guide/open-distro/all-in-one-deployment/all_in_one.html



# Manager installation
export JAVA_HOME=/usr/ && yum install -y curl unzip wget libcap && yum install -y java-11-openjdk-devel
rpm --import https://packages.wazuh.com/key/GPG-KEY-WAZUH
cat > /etc/yum.repos.d/wazuh.repo << EOF
[wazuh]
gpgcheck=1
gpgkey=https://packages.wazuh.com/key/GPG-KEY-WAZUH
enabled=1
name=EL-$releasever - Wazuh
baseurl=https://packages.wazuh.com/4.x/yum/
protect=1
EOF
yum install -y wazuh-manager
systemctl daemon-reload
systemctl enable wazuh-manager
systemctl start wazuh-manager


# Filebeat installation
yum install -y filebeat
curl -so /etc/filebeat/filebeat.yml https://raw.githubusercontent.com/wazuh/wazuh-documentation/4.1/resources/open-distro/filebeat/7.x/filebeat_all_in_one.yml
curl -so /etc/filebeat/wazuh-template.json https://raw.githubusercontent.com/wazuh/wazuh/4.1/extensions/elasticsearch/7.x/wazuh-template.json
chmod go+r /etc/filebeat/wazuh-template.json
curl -s https://packages.wazuh.com/4.x/filebeat/wazuh-filebeat-0.1.tar.gz | tar -xvz -C /usr/share/filebeat/module
mkdir /etc/filebeat/certs
cp /etc/elasticsearch/certs/root-ca.pem /etc/filebeat/certs/
mv /etc/elasticsearch/certs/filebeat* /etc/filebeat/certs/
systemctl daemon-reload
systemctl enable filebeat
systemctl start filebeat
filebeat test output


# Kibana installation
yum install -y opendistroforelasticsearch-kibana
curl -so /etc/kibana/kibana.yml https://raw.githubusercontent.com/wazuh/wazuh-documentation/4.1/resources/open-distro/kibana/7.x/kibana_all_in_one.yml
mkdir /usr/share/kibana/data
chown -R kibana:kibana /usr/share/kibana/data
cd /usr/share/kibana
sudo -u kibana /usr/share/kibana/bin/kibana-plugin install https://packages.wazuh.com/4.x/ui/kibana/wazuh_kibana-4.1.2_7.10.0-1.zip
mkdir /etc/kibana/certs
cp /etc/elasticsearch/certs/root-ca.pem /etc/kibana/certs/
mv /etc/elasticsearch/certs/kibana_http.key /etc/kibana/certs/kibana.key
mv /etc/elasticsearch/certs/kibana_http.pem /etc/kibana/certs/kibana.pem
setcap 'cap_net_bind_service=+ep' /usr/share/kibana/node/bin/node
systemctl daemon-reload
systemctl enable kibana
systemctl start kibana
```

**Comment by alberpilot** [comment]
PoC done, no technical limitations were found.
